src/Groups.jl

```diff
@@ -4,4 +4,265 @@
 #
 ###############################################################################
 
-elem_type(T::Type{<:Group}) = eltype(T)
+###############################################################################
+#
+#   Fundamental operations to be implemented for all group types
+#
+###############################################################################
+
+Base.one(G::Group) = throw(NotImplementedError(:parent, g))
+
+Base.parent(g::GroupElem) = throw(NotImplementedError(:parent, g))
+
+Base.:*(g::T, h::T) where {T<:GroupElem} = throw(NotImplementedError(:*, g, h))
+
+Base.inv(g::GroupElem) = throw(NotImplementedError(:inv, g))
+
+
+# also should implement:
+#   parent_type(::Type{MyGroupElem})
+#   elem_type(::Type{MyGroup})
+#
+# ... and many more of course (among those below and beyond), such as
+#   gens(::Group)
+#   ngens(::Group)
+#   order(::Group)
+#   order(::GroupElem)
+
+
+###############################################################################
+#
+#   Orders of groups and group elements
+#
+###############################################################################
+
+struct InfiniteOrderError{T} <: Exception
+    x::T
+    InfiniteOrderError(g::Union{GroupElem, Group}) = new{typeof(g)}(g)
+end
+
+function Base.showerror(io::IO, err::InfiniteOrderError{T}) where T
+    println(io, "Infinite order exception with ", err.x)
+    print(io, "order will only return a value when it is finite. ")
+    f = if T <: Group
+        "is_finite(G)"
+    elseif T <: GroupElem
+        "is_finiteorder(g)"
+    end
+    print(io, "You should check with `$f` first.")
+end
+
+@doc raw"""
+    order(::Type{T} = BigInt, G::Group) where T
+
+Return the order of $G$ as an instance of `T`.
+If $G$ is of infinite order, an `InfiniteOrderError` exception will be thrown.
+Use `is_finite(G)` to avoid this kind of exception.
+If the order does not fit into type `T`, an `InexactError` exception will be thrown.
+"""
+function order(::Type{T}, G::Group) where T
+    throw(NotImplementedError(:order, G))
+end
+
+@doc raw"""
+    order(::Type{T} = BigInt, g::GroupElem) where T
+
+Return the order of $g$ as an instance of `T`.
+If $g$ is of infinite order, an `InfiniteOrderError` exception will be thrown.
+Use `is_finiteorder(G)` to avoid this kind of exception.
+If the order does not fit into type `T`, an `InexactError` exception will be thrown.
+"""
+function order(::Type{T}, g::GroupElem) where T
+    throw(NotImplementedError(:order, g))
+end
+
+# if no return type has been specified, default to `BigInt`
+order(G::Group) = order(BigInt, G)
+order(g::GroupElem) = order(BigInt, g)
+
+"""
+    is_finiteorder(g::GroupElem)
+
+Return `true` if `g` is of finite order, possibly without computing it.
+"""
+function is_finiteorder(g::GroupElem)
+    isfinite(parent(g)) && return true
+    throw(
+        InterfaceNotImplemented(
```

Comment:
This should be `NotImplementedError` with different args

Reply:
done